models/multiphase/d2q9_ShanChen/Dynamics.R

```diff
@@ -0,0 +1,48 @@
+
+# Fluid Density Populations
+AddDensity( name="f[0]", dx= 0, dy= 0, group="f")
+AddDensity( name="f[1]", dx= 1, dy= 0, group="f")
+AddDensity( name="f[2]", dx= 0, dy= 1, group="f")
+AddDensity( name="f[3]", dx=-1, dy= 0, group="f")
+AddDensity( name="f[4]", dx= 0, dy=-1, group="f")
+AddDensity( name="f[5]", dx= 1, dy= 1, group="f")
+AddDensity( name="f[6]", dx=-1, dy= 1, group="f")
+AddDensity( name="f[7]", dx=-1, dy=-1, group="f")
+AddDensity( name="f[8]", dx= 1, dy=-1, group="f")
+
+AddDensity( name="rho", dx=0, dy=0, group="density") # is it the same as  AddField("rho", dx=c(0,0), dy=c(0,0), group="density")
```

Comment:
Q: Is this the same for all cases?
AddField("xa", dx=c(0,0), dy=c(0,0), group="my_xa") 
// or 
AddDensity( name="xb", dx=0, dy=0, group="my_xb")

then
```c
real_t var1 = xa;
// or
real_t var2 = xa(0,0);

```

Reply:
This is not a simple question :)
Densities gets 'advected', are persistent and rw. Fields are "compute-synchronize-allow neighbor access" type of variable. At least as far as I know.

For Shan-Chen gradient evaluation You most likely want Field.
Hope it helps.

Reply:
by 'advected' means that one can write to other density cell?
for example 
my_density[1] = 123; // ok
my_field(0,1) = 123; // shall crash?

however
real_t var1 = myfield; // ok
real_t var2 = myfield(0,0); // ok - same as var1

myfield(0,0) =123; // ok? or is it an read access pattern only?
myfield = 234; //ok

Reply:
While having `AddDensity( name="rho", dx=0, dy=0, group="density")`
```c
CudaDeviceFunction real_t getRho() {
    // This function defines the macroscopic density at the current node.
    return rho; // works fine, and Rho is in the output
}
```